만국박람회 [STEP 2] 재재, 보리사랑

ZZBAE opened the pull request:
@ehgud0670  제이슨 안녕하세요 😊 
만국박람회 step2 함께 진행 후 PR 보냅니다. 리뷰 잘 부탁드립니다 🫡


# 만국박람회 Step 2
> 프로젝트 기간 : 2022.06.15 ~ 2022.06.18
> 팀원 : 재재, 보리사랑
> 리뷰어 : 제이슨

## 개발환경 및 라이브러리
[![swift](https://img.shields.io/badge/swift-5.6-orange)]() [![xcode](https://img.shields.io/badge/Xcode-13.3.1-blue)]()

## Step 2 수행 중 핵심경험
- [x] 테이블뷰의 Delegate와 Data Source의 역할의 이해
- [x] 테이블뷰의 셀의 재사용 이해
- [x] 테이블뷰의 전반적인 동작 방식의 이해
- [x] 주어진 JSON 데이터를 파싱하여 테이블뷰에 표시
- [x] 내비게이션 컨트롤러를 활용한 화면 전환
- [x] 뷰 컨트롤러 사이의 데이터 전달


## TroubleShooting
1. **Navigation의 BarItem 변경시 view life cycle마다 ViewController가 다르게 인식**
네비게이션 바의 타이틀을 변경하기 위해 다음과 같은 메서드를 구현해보았습니다.
```swift
func modifyNavigationTitle(to name: String, in VC: UIViewController) {
    VC.title = name
}
```
그리고 이 메서드를 `viewWillAppear`에서 사용하였습니다
```swift
override func viewWillAppear(_ animated: Bool) {
    modifyNavigationTitle(to: "한국의 출품작", in: self)
    modifyNavigationBackButtonTitle(to: "메인", in: self)
}
```
다만 이 메서드를 viewWillAppear와 viewDidLoad 두군데서 실행 시켜 보았는데. self가 이전 뷰컨트롤러와 현재 뷰컨트롤러로 다르게 나오는걸 확인했습니다. 
이 부분도 `VC.title`이 아니라 `self.navigationController?.navigationBar.topItem?.title` 과 같이 정확하게 지정하여 변경하고자 하는데 view lifecycle과 self의 연관성 확인 후 변경할 예정입니다.


2. **Description의 줄바꿈이 되지 않고 한 줄로만 나오는 경우**
JSON에서 파싱한 description data의 줄바꿈을 해주려고 하였지만, 우선 구현에 목적을 두고 refactor 혹은 step3에서 오토레이아웃을 적용시켜 바꿔줄 계획입니다.



## 조언을 구하고 싶은 부분
1. **JSONData Decoding 메서드의 재사용성 증가 관련 궁금증**
기존에 STEP1에서는 각 테스트 별로 디코딩하는 메서드를 생성해서 사용했지만 이번 STEP2에서는 이 메서드의 재사용성을 높여주고자 전역 메서드로 분리를 시켰습니다. 여러 타입을 같은 메서드에서 사용하고자 할 때 프로토콜로 묶어서 쓴 기억이 계산기 떄 있었기에 이번에도 `ExpoData`라는 프로토콜을 만들었습니다. 
`ExpoHomeViewController`가 사용할 타입인 `Exposition`, `ExpoMenuVuewController`에서 사용될 `ExpoItemElement`배열 타입을 위해 각각의 viewDidLoad 메서드에서 이를 호출하게 했습니다.

```swift
func loadJsonData<ExpoData: Decodable>(type: ExpoData.Type ,_ fileName: String, _ extensionName: String) -> ExpoData {
    let jsonDecoder = JSONDecoder()
    let fileUrl = loadFileLocation(fileName, extensionName)
    do {
        let data = try Data(contentsOf: fileUrl)
        let decodedData = try jsonDecoder.decode(type.self, from: data)
        return decodedData
    } catch {
        fatalError("JSON DATA LOAD ERROR: \(fileName),\(extensionName)")
    }
}
```
다만 작성하다보니 JSONDecoder의 `decode`메서드에 들어가는 argumentLabel이 메타타입이었고 프로토콜을 메타타입으로 지정하려고하니 제너릭으로 사용해야 한다는 오류가 있었습니다. 
정확하게 어떻게 동작하는지 이해는 하지 못하였고 제너릭을 막 붙여보니 잘 동작해서 일단 사용하였는데 이런식으로 사용하는게 맞는 방향인지 궁금해서 질문드립니다.

2. **뷰 간의 데이터 이동**
뷰와 뷰 사이의 데이터 이동 / 주입을 시킬 때, 밑의 메서드를 활용하여
```swift
instantiateViewController(with Identifier:)
```
컨트롤러에 프로퍼티를 만들어서 넘겨주는 방법을 사용하였습니다. 이 외에 다른 간단한 방법이 있는지 궁금합니다!



---
## 커밋 규칙
* feat    : 기능 추가 (새로운 기능)
* fix      : 버그 수정
* refactor : 리팩토링 (네이밍 수정 등)
* style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
* docs    : 문서 변경 (문서 추가, 수정, 삭제)
* test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)

## 코드 컨벤션
Swift Statements 형식 준수

ZZBAE committed:
feat: 첫 화면 View Controller 생성

- Scroll View, Stack View 등을 생성하여 label, image, button 구성을 해주었습니다.
- 각각의 Constraints와 layout을 맞춰주었습니다.

yusw10 committed:
docs: Updatge README.md

yusw10 committed:
feat: ExpoHomeViewController에 IBOutlet 및 IBAction 설정

- Label 및 ImageView의 IBOutlet연결
- ExpoMenuViewController로 넘어가기 위한 pushToExpoMenuButtonDidTap IBAction 등록

yusw10 committed:
feat: exposition.json파일을 디코딩하는 메서드 구현 및 적용

- loadJsonData 메서드를 통해 exposition데이터 업데이트
- loadFileLocation메서드를 통해 타겟 json파일 경로 획득

ZZBAE committed:
feat: UIUpdate 메서드 구현

- Exposition의 케이스들을 각 label에 데이터로 할당하여 ㄸxpoHomeViewController에 보여지도록 업데이트
- Assets에서의 이미지 이름으로 UIImage 데이터 가져옴
- visitors의 Int값을 String으로 변환

yusw10 committed:
feat: 화면 전환을 위한 pushToExpoMenuButtonDidTap 메서드 구현

- navigationController를 활용하여 push 방식으로 ExpoMenuViewController로 화면전환하는 기능 추가

yusw10 committed:
feat: ExpoMenuViewController 추가

- storyboard에 ExpoMenuViewController 추가

ZZBAE committed:
feat: ExpoMenuViewController에 오토레이아웃 설정

- storyboard에서 ExpoMenuViewController에 해당하는 Image, title, description 레이아웃 설정

yusw10 committed:
refactor: ExpoData프로토콜 구현 및 Json관련 메서드 분리

- Exposition, ExpoItemElement가 ExpoData 프로토콜을 준수하도록 변경
- loadJsonData, loadFileLocation 메서드를 재사용하기 위해 리팩토링
- ExpoHomeViewController에서 Json 데이터를 디코딩 하는 메서드를 분리

yusw10 committed:
feat: ExpoHomeViewController에서 ExpoItemElement 데이터 디코딩

- Table View  컨텐츠의 데이터가 될 ExpoItemList를 뷰컨트롤러의 전역으로 선언
- "items.json" 파일로부터 ExpoItemElement의 배열 인스턴스 디코딩 및 저장

ZZBAE committed:
feat: ExpoMenuTableViewCell에 IBOutlet 연결 및 UI Update

- ExpoMenuTableViewCell에 ExpoItemImage, Title, Description를 storyboard와 IBOutlet 연결
- settingImageView, settingLabelView 메서드 생성
- Cell 데이터 업데이트를 하는 메서드 생성

ZZBAE committed:
feat: ExpoMenuTableViewCell에 IBOutlet 연결

- ExpoMenuTableViewCell에 ExpoItemImage, Title, Description를 storyboard와 IBOutlet 연결
- settingImageView, settingLabelView 메서드 생성

yusw10 committed:
Merge branch 'step2' of https://github.com/ZZBAE/ios-exposition-universelle into step2

yusw10 committed:
feat: ExpoMenuViewController의 TableView 기능 구현

- expopItemTableView의 DataSource와 Delegate구현 및 기능 구현을 ExpoMenuViewContgroller에 위임
- 셀 선택시 세부 화면으로 넘어가기 위한 didSelectRowAt 메서드 구현
- DetailViewController 구현 및 storyboard에 뷰 추가

yusw10 committed:
refactor: ExpoMenuTableViewCell에 연결된 IBOutlet 프로퍼티 수정

- settingLableView의 argument label 및 프로퍼티 변경 코드 수정
- ExpoItemTitleLabel에서 ExpoItemNameLabel로 프로퍼티 명 변경

ZZBAE committed:
feat: DetailViewController 구현

- ExpoMenuViewController에서 DetailViewController로 데이터 주입
- storyboard에 보이도록 updateUI 메서드 생성
- storyboard와 DetailViewController IBOutlet 연결

yusw10 committed:
style: 코드 작성자 수정 및 공백, 주석 제거

yusw10 committed:
refactor: ExpoHomeViewController의 버튼 title 수정

yusw10 committed:
feat: Navigation Bar의 TItle과 backbutton title 수정

- modifyNavigationTitle 메서드를 구현하여 navigation bar의 title이 변경되도록 수정
- modifyNavigationBackButtonTitle 메서드를 구현하여 뒤로가기 버튼의 title 수정

ehgud0670 commented on Expo1900/Expo1900/Util/JsonDataDecoding.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Foundation
+
+func loadJsonData<ExpoData: Decodable>(type: ExpoData.Type ,_ fileName: String, _ extensionName: String) -> ExpoData {
```
> 조언을 구하고 싶은 부분

저라면 프로토콜 ExpoData 을 채택한 모델 말고도 다른 모델도 해당 메소드를 사용하기 위해 ExpoData라는 건 사용하지 않을 것 같아요. 저번 계산기 미션은 큐에서 들어갈 Element를 구분하기 위해 빈 프로토콜을 사용한 것 같은데, 
지금 jsonData를 처리하는건 모든 모델에 대해 처리할 수 있는 것 같아 ExpoData라는 프로토콜은 사용하지 않을 것 같습니다. 
이미 모델은 Codable(or Decodable)을 채택하고 있기 때문입니다 :)

ZZBAE replied:
제이슨이 말씀해주신 것 처럼 이미 Codable을 채택하고 있기 때문에 기존의 ExpoData 프로토콜을 사용하지 않아 파일 자체로 제거해주었습니다. 그리고 ExpoItemElement, Exposition 에서 채택하고 있는 것도 수정해주었습니다.

[ExpoData 프로토콜 채택 제거](https://github.com/yagom-academy/ios-exposition-universelle/pull/183/commits/f8057efc9f145d496d7321d7fe101a6fb7284fb0)

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoHomeViewController.swift:
```diff
@@ -0,0 +1,49 @@
+        self.navigationController?.pushViewController(menuVC, animated: true)
+    }
+    
+    private func setExpositionData() {
```
설정하는 메소드로 네이밍은 set 보단 setup, configure가 어울립니다.
현재 set은 get,set 메소드와 이름이 겹치기 하고요

ZZBAE replied:
지적해주신대로 setup으로 통일하여 메서드 네임 변경해주었습니다!

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoMenuViewController.swift:
```diff
@@ -0,0 +1,57 @@
+        setExpositionData()
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        modifyNavigationTitle(to: "한국의 출품작", in: self)
+        modifyNavigationBackButtonTitle(to: "메인", in: self)
+    }
```
1) super 메소드가 빠졌습니다!
2)viewDidLoad에서 같이 해줘도 되는 처리같은데 viewWillAppear에서 할 이유가 있었을까요~?

ZZBAE replied:
저희가 테스트 실행을 하였을 때, viewDidLoad에서 같이 해주니까 NavigationTitle이 그 전 viewController로 밀리는 현상들이 발견되어 viewWillAppear로 따로 해주었습니다. 근데 지금 다시 viewDidLoad에서 같이 해주니까 밀림 현상이 없어서 당황스럽네요... 이유를 못찾겠는데 혹시 제이슨은 아시나요?

ehgud0670 replied:
추측해보기론 시뮬레이터로 돌리면 밀림현상이 발생하는 것 같습니다. 한번 디바이스로도 테스트하셔도 밀림현상이 발생하면 viewWillAppear에 해야 할텐데 아니라면 viewDidAppear에서 해주시면 될 것 같아요 :)

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoMenuViewController.swift:
```diff
@@ -0,0 +1,57 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: "ExpoMenuTableViewCell") as? ExpoMenuTableViewCell else {
```
reuseIdentifier를 편하게 관리할 수 있도록 직접 프로토콜을 만들어 사용하실 수도 있습니다. 
참고차 말씀드려요~

link: https://medium.com/bleeding-edge/nicer-reuse-identifiers-with-protocols-in-swift-97d18de1b2df

ehgud0670 reviewed (commented):
여기 PR에는 테스트코드가 보이지 않지만 테스트코드를 보니 제 피드백이 반영안된 것 같네요!
디코딩한 객체가 잘 디코딩됐는지 확인하는 것이 디코딩 테스트지, 
서로 다른 객체와 비교하면서 디코딩이 잘 됐는지 테스트하는 것은 제가 보기엔 아닌 테스트같아요. 그럼 다시 한번 고민해보세요 :)

더 피드백 드릴것은 없나 다시 코드를 보도록 하겠습니다.

ehgud0670 commented on Expo1900/Expo1900/Util/NavigationItemManager.swift:
```diff
@@ -0,0 +1,16 @@
+//
+//  NavigationItemManager.swift
+//  Expo1900
+//
+//  Created by 재재, 보리사랑 on 2022/06/18.
+//
+
+import UIKit
+
+func modifyNavigationBackButtonTitle(to name: String, in VC: UIViewController) {
+    VC.navigationController?.navigationBar.topItem?.backButtonTitle = name
+}
+
+func modifyNavigationTitle(to name: String, in VC: UIViewController) {
+    VC.title = name
+}
```
음, uiviewcontroller extension을 이용해 메소드를 만들면 괜찮을 것 같아요 :)

ehgud0670 replied:
아시겠지만 모든 변수명은 소문자로 시작하는 것이 좋습니다.

ZZBAE replied:
UIViewController extension으로 메서드 생성해주었습니다!

그리고 모든 변수명 소문자로 변경해주었습니다. 앞으로 간단한 네이밍 실수 유의하겠습니다 ㅠㅠ

ehgud0670 commented on Expo1900/Expo1900/View/ExpoMenuTableViewCell.swift:
```diff
@@ -0,0 +1,23 @@
+    }
+    
+    func settingLableView(name: String, shortDescription: String) {
+        self.ExpoItemNameLabel.text = name
+        self.ExpoItemDescriptionLabel.text = shortDescription
```
여기도 소문자로 네이밍하는게 좋을 것 같아요. 특별한 이유가 있었을까요~?

ZZBAE replied:
특별한 이유는 없지만, ViewController네이밍을 다 대문자로 시작해서 무의식적으로 Expo를 적어버린 것 같습니다. 변경해주었습니다!..🥲

ehgud0670 commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,32 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        updateUI()
+        modifyNavigationTitle(to: expoItemElement?.name ?? "", in: self)
+        modifyNavigationBackButtonTitle(to: "한국의 출품작", in: self)
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        modifyNavigationTitle(to: expoItemElement?.name ?? "", in: self)
+        modifyNavigationBackButtonTitle(to: "한국의 출품작", in: self)
+    }
```
viewWillAppear에서 안해줘도 된다면 코드 지워도 될 것 같아요 :)
이렇게 두번 setup한 이유가 있었을까요~?

ZZBAE replied:
두 번 setup한 이유는 없고, 저희의 실수입니다.. viewWillAppear에서 해주지 않아 삭제해주었습니다.

ehgud0670 commented on Expo1900/Expo1900/Assets.xcassets/bangjja.imageset/Contents.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "images" : [
+    {
+      "filename" : "bangjja~universal@1x.png",
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "filename" : "bangjja~universal@2x.png",
+      "idiom" : "universal",
+      "scale" : "2x"
+    },
+    {
+      "filename" : "bangjja~universal@3x.png",
+      "idiom" : "universal",
+      "scale" : "3x"
```
1x, 2x, 3x 차이가 무엇인지 아는 것도 중요합니다!
잘 모르신다면 찾아보시면 좋을 것 같아요 👍

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoMenuViewController.swift:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+extension ExpoMenuViewController: UITableViewDelegate {
```
채택한 프로토콜 당 이렇게 extension으로 나눠주신거 잘하셨습니다! 👍👍👍
이러면 어떤 메소드(ex delegate 관련된 메소드)인지 파악하기 좋더라고요.

ZZBAE replied:
감사합니다! 🥰

ehgud0670 reviewed (commented):
많은 걸 배우시고 적용하셨네요 :)
수고 많으셨습니다. 그럼 피드백 확인 부탁드리고 의문이 들거나 궁금한 점 있으시면 말씀주세요!

ehgud0670 commented:
> 조언을 구하고 싶은 부분 

다른 방법으로 vc 을 만드는 건 아래 코멘트를 드리면 될 것 같아요 :) 
custom init을 사용해야 할 때 좋습니다.

https://github.com/yagom-academy/ios-juice-maker/pull/244#issuecomment-1126654925

ehgud0670 commented on Expo1900/Expo1900/View/ExpoMenuTableViewCell.swift:
```diff
@@ -0,0 +1,23 @@
+    @IBOutlet weak var ExpoItemNameLabel: UILabel!
+    @IBOutlet weak var ExpoItemDescriptionLabel: UILabel!
+    
+    func settingImageView(name: String) {
+        self.ExpoItemImageView.image = UIImage(named: name)
+    }
+    
+    func settingLableView(name: String, shortDescription: String) {
```
아시겠지만 여기 부분도 함수이므로 동사로 setup이나 configure가 네이밍이 더 좋을 것 같아요.

ZZBAE replied:
지적해주신대로 setup으로 통일하여 메서드 네임 변경해주었습니다!

ehgud0670 replied:
앗,, 지적은 아니고 첨언입니다 :)
두 분 모두 잘해주고 계세요! 화이팅! 💪🔥

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoMenuViewController.swift:
```diff
@@ -0,0 +1,57 @@
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: "ExpoMenuTableViewCell") as? ExpoMenuTableViewCell else {
+            return UITableViewCell()
+        }
+        let currentExpoItemElement = expoItemList[indexPath.row]
```
swift에서 크래시가 나는 대표적인 요인 중 하나는 없는 인덱스에 잘못 접근하는 것입니다. 
따라서 크래시가 애초에 발생하지 않도록, optional을 이용해 index에 safe하게 접근하는 방법이 있는데 참고차 공유드립니다!
swift에서 따로 제공하는 것이 아니고 extension을 이용해 직접 구현해주는 방법입니다.
link: https://stackoverflow.com/a/30593673

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoHomeViewController.swift:
```diff
@@ -0,0 +1,49 @@
+    @IBOutlet weak var leftFlagImageView: UIImageView!
+    @IBOutlet weak var rightFlagImageView: UIImageView!
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setExpositionData()
+        updateUI()
+    }
+    
+    private func updateUI() {
+        let visitorsNumber = String(exposition?.visitors ?? 0)
+        titleLabel.text = exposition?.title
+        mainImageView.image = UIImage(named: "poster")
+        visitorsLabel.text = visitorsNumber
+        locationLabel.text = exposition?.location
+        durationLabel.text = exposition?.duration
+        descriptionLabel.text = exposition?.expositionDescription
+        leftFlagImageView.image = UIImage(named: "flag")
+        rightFlagImageView.image = UIImage(named: "flag")
+    }
+    
+    @IBAction func pushToExpoMenuButtonDidTap(_ sender: UIButton) {
+        guard let menuVC = storyboard?.instantiateViewController(withIdentifier: "ExpoMenuViewController") as? ExpoMenuViewController else { return }
+        self.navigationController?.pushViewController(menuVC, animated: true)
+    }
+    
+    private func setExpositionData() {
+        let exposition = loadJsonData(type: Exposition.self, "exposition", "json") 
+        self.exposition = exposition
+    }
```
```suggestion
    override func viewDidLoad() {
        super.viewDidLoad()
        setExpositionData()
        updateUI()
    }
    
    private func setExpositionData() {
        let exposition = loadJsonData(type: Exposition.self, "exposition", "json") 
        self.exposition = exposition
    }
    
    private func updateUI() {
        let visitorsNumber = String(exposition?.visitors ?? 0)
        titleLabel.text = exposition?.title
        mainImageView.image = UIImage(named: "poster")
        visitorsLabel.text = visitorsNumber
        locationLabel.text = exposition?.location
        durationLabel.text = exposition?.duration
        descriptionLabel.text = exposition?.expositionDescription
        leftFlagImageView.image = UIImage(named: "flag")
        rightFlagImageView.image = UIImage(named: "flag")
    }
    
    @IBAction func pushToExpoMenuButtonDidTap(_ sender: UIButton) {
        guard let menuVC = storyboard?.instantiateViewController(withIdentifier: "ExpoMenuViewController") as? ExpoMenuViewController else { return }
        self.navigationController?.pushViewController(menuVC, animated: true)
    }
```

=> 함수 호출 순으로 메소드를 위치하는 것이 책 클린코드에서 알려주는 기본적인 스타일입니다. 
물론 이 스타일도 예외는 존재하지만 가장 많이 따르는 스타일이라 참고차 말씀드립니다 :)

ehgud0670 commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,32 @@
+
+import UIKit
+
+class DetailViewController: UIViewController {
```
```suggestion
final class DetailViewController: UIViewController {
```
=> final 을 붙이면 성능적으로 좀 더 효율적입니다. 필수는 아니고 참고차 말씀 드립니다. 
지금 단계에서는 모르셔도 되는 부분이니 그런가보다 하고 넘어가시면 될 듯 합니다 :)

> 참고 자료

* [swift-optimizationTips](https://github.com/apple/swift/blob/main/docs/OptimizationTips.rst#advice-use-final-when-you-know-the-declaration-does-not-need-to-be-overridden)
* [스위프트 성능 이해하기-유용하](https://www.youtube.com/watch?v=z1Gf6EosaUQ)

yusw10 committed:
refactor: ExpoData 프로토콜 채택 제거

- ExpoData 프로토콜 제거
- ExpoItemElement, Exposition 데이터 타입에서 ExpoData프로토콜 채택 제거
- loadJsonData 메서드의 전달인자 타입을 ExpoData프로토콜을 상속받는 타입에서 Decodable을 상속받는 generic 타입을 받도록 변경

ZZBAE committed:
test: decoding을 확인할 수 있는 test case 수정

- 기존의 test 에서 URL을 찾지 못하여 tesetBundle 활용하여 변경해줌

yusw10 committed:
refactor: 네비게이션 바의 text 업데이트 메서드 수정

- modifyNavigationTitle과 modifyNavigationBackButtonTitle 메서드를 UIviewController의 extension으로 확장
- modifyNavigationTitle과 modifyNavigationBackButtonTitle 메서드를 viewDidload에서 호출하도록 변경
- DetailViewController의 viewWillAppear 메서드 삭제

yusw10 committed:
refactor: 네비게이션 바의 text 업데이트 메서드 수정

- modifyNavigationTitle과 modifyNavigationBackButtonTitle 메서드를 UIviewController의 extension으로 확장
- modifyNavigationTitle과 modifyNavigationBackButtonTitle 메서드를 viewDidload에서 호출하도록 변경

yusw10 committed:
Merge branch 'step2' of https://github.com/ZZBAE/ios-exposition-universelle into step2

ZZBAE committed:
refactor: ExpoHomeVeiwController viewDidLoad 구현 순서 변경

Co-authored-by: Doe-Hyung KIM <kimdo2297@gmail.com>

ZZBAE committed:
refactor: DetailViewController 리뷰 반영 final 추가

Co-authored-by: Doe-Hyung KIM <kimdo2297@gmail.com>

yusw10 committed:
refactor: 메서드명 및 프로퍼티명 수정

- setExpositionData -> seupExpositionData 로 메서드명 변경
- settingImageView -> setupImageView 로 메서드명 변경
- settingLabelView -> setupLabelView로 메서드명 변경
- modifyNavigationBackButtonTitle 및 modifyNavigationTitle의 전달인자 VC를 viewController로 변경

yusw10 committed:
Merge branch 'step2' of https://github.com/ZZBAE/ios-exposition-universelle into step2

# Conflicts:
#	Expo1900/Expo1900/Controller/ExpoHomeViewController.swift

yusw10 committed:
refactor: 메서드명 및 프로퍼티명 수정

- setExpositionData -> setupExpositionData 로 메서드명 변경
- settingImageView -> setupImageView 로 메서드명 변경

ehgud0670 commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,27 @@
+final class DetailViewController: UIViewController {
+    var expoItemElement: ExpoItemElement?
+    
+    @IBOutlet weak var ExpoItemImageView: UIImageView!
+    @IBOutlet weak var ExpoItemDescriptionLabel: UILabel!
```
여기도 고쳐주시면 좋을 것 같아요 :)

ehgud0670 commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,27 @@
+import UIKit
+
+final class DetailViewController: UIViewController {
+    var expoItemElement: ExpoItemElement?
```
private var 로 바꾸고 제가 말씀드린 custom init 을 사용해도 좋을 듯 합니다. 참고차 말씀드려요 :)
https://github.com/yagom-academy/ios-exposition-universelle/pull/183#issuecomment-1159667592

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoMenuViewController.swift:
```diff
@@ -0,0 +1,54 @@
+import UIKit
+
+class ExpoMenuViewController: UIViewController {
+    var expoItemList: [ExpoItemElement] = []
```
배열을 따로 관리하는 모델 객체(네이밍은 예를 들어 `ExpoItems`)를 따로 만들고 동작(메소드 or 연산 프로퍼티)으로 표현해도 좋을 것 같아요. 
참고차 말씀드립니다 :)

ehgud0670 commented on Expo1900/Expo1900/View/ExpoMenuTableViewCell.swift:
```diff
@@ -0,0 +1,23 @@
+    @IBOutlet weak var expoItemNameLabel: UILabel!
+    @IBOutlet weak var expoItemDescriptionLabel: UILabel!
+    
+    func setupImageView(name: String) {
+        self.expoItemImageView.image = UIImage(named: name)
+    }
+    
+    func setupLableView(name: String, shortDescription: String) {
+        self.expoItemNameLabel.text = name
+        self.expoItemDescriptionLabel.text = shortDescription
+    }
```
```suggestion
    func setupUI(with expoItemElement: ExpoItemElement) {
         setupLableView(name: expoItemElement.name, shortDescription: expoItemElement.shortDesc)
         setupImageView(name: expoItemElement.imageName)
    } 

    private func setupImageView(name: String) {
        self.expoItemImageView.image = UIImage(named: name)
    }
       
    private func setupLableView(name: String, shortDescription: String) {
        self.expoItemNameLabel.text = name
        self.expoItemDescriptionLabel.text = shortDescription
    }
```
=> 이렇게 모델 값을 전달 인수로 넘겨 외부에서 한번만(setupUI만) 호출하는 방법도 있습니다. 이러면 외부에서는 setupUI(with:)만 호출하고 cell 코드 안쪽에서만 코드가 변경하면 되는 장점이 있습니다.

ehgud0670 commented on Expo1900/Expo1900/View/ExpoMenuTableViewCell.swift:
```diff
@@ -0,0 +1,23 @@
+    @IBOutlet weak var expoItemImageView: UIImageView!
+    @IBOutlet weak var expoItemNameLabel: UILabel!
+    @IBOutlet weak var expoItemDescriptionLabel: UILabel!
+    
```
cell 의 prepareForReuse라는 메소드가 있는데 꼭 알아두셨으면 좋겠습니다.

ehgud0670 reviewed (approved):
고생하셨습니다 :)
